### PR TITLE
Pythonise, test, and refactor cpdFindChangePoint function

### PR DESCRIPTION
**Pull request by casterkay**
## Summary
This pull request extends the unit tests for the cpdFindChangePoint MATLAB function. The cpdFindChangePoint20100901 function is a MATLAB function designed to detect change points in a given dataset. Here's a detailed breakdown of what the function does:

### Functionality

1. The function initializes the output variables Cp2, Cp3, s2, and s3 to NaN or empty structures.
2. Data Preprocessing:
    - Reshapes the input vectors xx and yy.
    - Excludes missing data points.
    - Excludes extreme linear regression outliers.
3. Computes the statistics of reduced (null hypothesis) models for later testing of Cp2 and Cp3 significance.
4. Computes the F score for each data point to identify the maximum F score (Fmax) for both the 2-parameter and 3-parameter models.
5. Assigns change points from the maximum F scores and calculates statistics.
6. Tests for the significance of the Fmax scores.
7. If fPlot is set to 1, the function plots the data points, fitted models, and identified change points.

### Key Points

- The function uses linear regression to fit models and identify change points.
- It handles missing data and outliers to ensure robust change-point detection.
- The function provides detailed statistics for the identified change points, including confidence intervals and significance levels.
- The change points are set to NaN if they are not statistically significant.

## Changes Made
- Extended Test Cases:
  - Added tests for scenarios with no change points.
  - Added tests for change points at the start and end of the data.
  - Added tests for data with noise.
  - Added tests for invalid inputs and insufficient data.
- Test Files Modified:
  - test_cpdFindChangePoint20100901.py: Extended with new test cases to cover more edge cases and scenarios.#
 
## Detailed Description
- No Change Points: Added a test to verify the function's behavior when there are no change points in the data.
- Change Points at Start/End: Added tests to check if the function correctly identifies change points located at the start or end of the data series.
- Data with Noise: Added a test to ensure the function can still identify change points when the data contains noise.
- Invalid Inputs: Added tests to handle scenarios with invalid inputs, such as empty arrays or insufficient data points.

### CODE REVIEW

**Comment by dorchard** [comment]
Reviewing in progress. Could we get a quick summary of what `cpdFindChangePoint` does to help assess the tests? This could also be added to the MATLAB code so we have extra documentation.

**Comment by dorchard** [comment]
@isaacaka has some documentation too that can be included.

**Comment by j-emberton** [comment]
Hi @tztsai,

I've updated the branch we are merging into from `main` to `ustar_cp_refactor_main`. Because of this there now seems to be a conflict between the `tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py` files.

I've also tried to run the tests locally but get the following error:
image

I'll take a look at the tests separately just now but won't be able to run them.

**Comment by casterkay** [comment]
> Hi @tztsai,
> 
> I've updated the branch we are merging into from `main` to `ustar_cp_refactor_main`. Because of this there now seems to be a conflict between the `tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py` files.
> 
> I've also tried to run the tests locally but get the following error: image
> 
> I'll take a look at the tests separately just now but won't be able to run them.

It seems due to a conflict between `ustar_cp_refactor_main` and `main`, not related to the update here. Perhaps `context.py` should be removed from `ustar_cp_refactor_main`.

**Comment by j-emberton** [comment]
I don't think you have branched off of ustar_cp_refactor_main, therefore the rest of the code in your branch isn't up to date and you don't have access to all the code that we've developed to date. You should rebase your branch and make sure the tests work before requesting a review.

**Review by j-emberton — CHANGES_REQUESTED** [review]
This code isn't ready for review. Please fix and resubmit.

**Comment by casterkay** [comment]
I've rebased to the head of `ustar_cp_refactor_main` now.

**Review by j-emberton — DISMISSED** [review]
Not all tests pass when I run locally. This PR isn't ready for review.

**Comment by dorchard** [comment]
I can confirm that I've got tests failing too. It looks like a problem with the rebase:
```
============================================================== short test summary info ===============================================================
FAILED tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py::test_cpdFindChangePoint_no_change - AssertionError: Cp2 should be NaN when there's no change point
FAILED tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py::test_cpdFindChangePoint_change_at_start - AssertionError: Cp2 should be near the start
FAILED tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py::test_cpdFindChangePoint_change_at_end - AssertionError: Cp2 should be near the end
FAILED tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py::test_cpdFindChangePoint_with_noise - AssertionError: Cp2 should be near 5 despite noise
FAILED tests/unit_tests/test_ustar_cp/test_rounding.py::test_round - assert 3 == 2
====================================================== 5 failed, 16 passed in 74.17s (0:01:14) =======================================================
```

**Comment by j-emberton** [comment]
This PR is now up to date with `ustar_cp_refactor_main` and includes Issacs fix from #61 .

4 tests associated with the specific `cpdFindChangePoint` function now fail but all others pass

**Comment by dorchard** [comment]
Could do with some site-based data tests.

**Review by dorchard — APPROVED** [review]

File: tests/conftest.py [hunk 1]
```diff
@@ -131,10 +132,19 @@ class PythonEngine(TestEngine):
     def _repr_pretty_(self, *args):
         return "Python Test Engine"
 
+#'to_python' needs to be added back in
```
Comment (j-emberton) [inline tests/conftest.py]: Can we remove these old comments

File: tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py [hunk 2]
```diff
@@ -18,31 +100,305 @@ def test_cpdFindChangePoint20100901(test_engine, test_data):
...
+
+]
+@pytest.mark.parametrize('xx, yy, expected_xx, expected_yy', testcase_removeNans)
+def test_removeNans(test_engine, xx, yy, expected_xx, expected_yy):
+
+    # Call the MATLAB function
+    xx, yy = test_engine.removeNans(test_engine.convert(xx), test_engine.convert(yy), nargout=2)
+
+    # Perform assertions
+    # x = 1.1111111111111111111
```
Comment (j-emberton) [inline tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py]: I think these comments can be deleted now

File: tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py [hunk 3]
```diff
@@ -18,31 +100,305 @@ def test_cpdFindChangePoint20100901(test_engine, test_data):
...
+
+    n = len(xx)
+
+    # Call the MATLAB function
+    test_x, test_y = removeOutliers(xx, yy, n)
+    x, y = test_engine.removeOutliers(test_engine.convert(xx), test_engine.convert(yy), n, nargout=2)
+
+    # Perform assertions
+    # print(x[:10].shape, y[:10].shape)
+    # print(expected_x[:10].shape, expected_y[:10].shape)
```
Comment (j-emberton) [inline tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py]: I think these comments can be deleted

File: tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py [hunk 4]
```diff
@@ -18,31 +100,305 @@ def test_cpdFindChangePoint20100901(test_engine, test_data):
...
+    # assert False
+    print(x, y)
+    print(test_x, test_y)
+    assert test_engine.equal(test_engine.convert(x), test_engine.convert(expected_x))
+    assert test_engine.equal(test_engine.convert(y), test_engine.convert(expected_y))
+
+
+## TODO: Add site data tests for computeReducedModels
```
Comment (j-emberton) [inline tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py:225]: To do?

File: tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py [hunk 5]
```diff
@@ -18,31 +100,305 @@ def test_cpdFindChangePoint20100901(test_engine, test_data):
...
+    # python_fc3 = fitOperational3ParamModel(*input_data.values())
+
+    Fc3 = test_engine.fitOperational3ParamModel(*input_data.values(), nargout=1)
+
+    output_names = ['Fc3']
+    output_data = load_data(test_engine, artifacts_dir, year, iteration, output_names, 'CA-Cbo_qca_ustar', 'output')
+
+    print("FC£: ",Fc3)
+    # print(python_fc3)
+    #print(output_data['Fc3'])
```
Comment (j-emberton) [inline tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py]: These comments can be deleted

**Comment by j-emberton** [comment]
Good work.

I think there's just a few out of date comments that can be deleted to tidy up then this is ready for merging.

File: tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py (hunk 4 above)
Comment (isaacaka) [reply tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py:225]: Yes, added that thanks.
